### ISSUE TITLE
False-positive IncompatibleMethodOverride when decorating an inherited method but not the original method

### ISSUE THREAD

**Issue by jshumphrey** [microsoft/pyright#10002]
## Describe the bug
When a subclass overrides a method with a parameter, the subclass decorates its override, and the parent class's original method is not decorated, the following error message is emitted, even though the decorator does not change the function signature at all.

> Method "my_method" overrides class "BaseClass" in an incompatible manner
  Parameter 2 mismatch: base parameter "my_param" is keyword parameter, override parameter is position-only  (reportIncompatibleMethodOverride)

### Context
I have an abstract base class with an abstract method. All subclasses of the base class decorate their implementations of that abstract method with a particular decorator, which executes a number of boilerplate error-checking steps before executing the method normally.

Because the original method from the abstract base class is meaningless (its body is simply `raise NotImplementedError`), I did not bother decorating it with the error-checking decorator. I eventually found that this was causing the IncompatibleMethodOverride errors I was seeing on all of the implementations in all of the subclasses of the base class.

I am surprised that this causes an error, and believe it may be an actual bug, on the grounds that the decorator does not change the function signature whatsoever, and returns the same type it's provided. Nevertheless, it would not surprise me if the answer to this is "decorating a method override, but not the original method, is not valid", and I need to change my code.

However, I wanted to file the bug report anyway because the error message that Pylance reports in this case is very misleading; when trying to address the cause of the errors, I was scouring my code for any place where I accidentally made `my_param` a keyword argument, but could never find one. If this pattern really is not valid, it would be nice to have some kind of better-tailored error message explaining that if you decorate a method override, you must also similarly decorate the original method.

## Code or Screenshots

```python
import abc
from collections.abc import Callable
import functools
import typing

AnySubclass = typing.TypeVar("AnySubclass", bound = "BaseClass")

def decorator(
    function: Callable[[AnySubclass, int], int]
) -> Callable[[AnySubclass, int], int]:
    """This decorator should be applied to the ``my_method`` method of every
    subclass of MyClass, to automate boilerplate error-checking steps."""
    
    @functools.wraps(function)  # Removing this decorator doesn't matter
    def wrapper(my_instance: AnySubclass, my_param: int) -> int:
        # Boilerplate error-checking, logging, and exception raising goes here
        # If no errors are found, just call the function normally
        return function(my_instance, my_param)
    return wrapper

class BaseClass(abc.ABC):
    #@decorator  # Uncommenting this decorator removes the issue
    @abc.abstractmethod
    def my_method(self, my_param: int) -> int:
        """Subclasses should always decorate this with ``decorator``."""
        raise NotImplementedError

class SubClass(BaseClass):
    @decorator  # Commenting this decorator removes the issue
    def my_method(self, my_param: int) -> int:  # Typing error on this line
        ...
```

Code sample in [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.344&strict=true&code=JYWwDg9gTgLgBAQwEYGMCwAoAZlCI4oQA2RApijMBAHYDOAdMinKJLHAMIInJmavR4WAK7UKEYrX7hBcGAE8wwagHNMmAILV5AZWGoiCWrTgBeOYuUr6AFUWkAagigAKAERbd%2BlIeNuANHBIEKIAJmZwbgBCRqQcvrRuAJTqGKGkWHDphFAIMNAumHDFcCJilDQAXJzchkhkANoNnnoGRrSByjAAup3UPZhJcAC0AHw1PPWkTS3eCX09C92VRSVu6zYAFsAm2dB50HC0myFE4UikiGBgRMCk4flym5cABi8g8gD6IKQwJ6FvOA-P4QcIQTKkABupCg8lWxVoc3acHBcAAsvJ4u1Ao8EMJ8iA8pdgsAyFAbkS4DDcFBhihnigANZWI4wUhgBjrNzwuA8gACZXEknoAHdchyXIKKtQhnAAMRwABKpBAEEhLL%2BOyy5H2%2BSgWQgpFo1AA5PBCTA2VAeelMmKENcYS4Pp9lLQYAgxKRqrM2sZAi6wM4ECBql0hmMWP0VhgSnH5XAohBSTCKWyqVAaXSGczVIEiBAVCorIFPeFSAAPFDs6VwXI7FkqQ0mZ5QUg8uMKgCSmWoEAzNJMzkuWBC1FCgQAVsJ3QRak8R6IKFRqHA%2B1BCSQ4bH48U2zBhFBV1KV86vm6PV6A18g7kQCkd3vfofV-bHdaMJgfMiYrQ4gkXCYegNCiDgkhjTs%2BT2XI9WKBUAFUxDwH5%2Bg1bZdh1GDDjbVVoRMP5Lh2RF20fOA%2BSA5B3VyChgX%2BG0MiBL5aNBFw-yILBr0%2BW8QzDfoI3GLoIN3SJ1lab9jCNI4TmEM5ECIEUEHkDCckpTUTBFYA-jgN5oIOKA3noLkOxKes-zgAA5CAYC7cAyBQtlQgAUUzaBUnEkxWixYwXF-f92nA-ldNghMOGQ0hUNUJ4tSC7CVTVSSCJYYxhBIuNbUY75fn%2BVjSHYzjuNDKMYH4orqgTOwlEi6lDhoKKTFuahUuE%2BgWqAA)

## VS Code extension or command-line
I'm running pyright as part of the Pylance VS Code extension, version 2025.2.1.

**Comment by erictraut** [comment]
The problem is that the decorator is changing the signature of the decorated method in a manner that makes it incompatible with the parent class. The method in the parent class indicates that it has one parameter  name `my_param` that accepts either positional or keyword arguments. In other words, you could call this method as `x.my_method(0)` or `x.my_method(my_param=0)`. The decorator produces a signature that accepts an argument by position only. It does not allow the caller to specify the argument by keyword. This is what pyright's error message is telling you.

You can fix this in your code one of two ways:

1. Modify the signature of the base class method to indicate that its parameter `my_param` is position-only. This is done by adding a `/` parameter to the signature. For details refer to [PEP 570](https://peps.python.org/pep-0570/).

```python
def my_method(self, my_param: int, /) -> int:
```

2. Modify the decorator to return a signature that allows the argument to be passed positionally or by keyword. This can't be done with the `Callable` special form. It requires the use of a [callback protocol](https://typing.readthedocs.io/en/latest/spec/callables.html#callback-protocols).

```python
class MyCallback[AnySubclass: BaseClass](typing.Protocol):
    def __call__(self: AnySubclass, my_param: int) -> int: ...


def decorator(function: Callable[[AnySubclass, int], int]) -> MyCallback[AnySubclass]: 
    ...
```

I'd probably go with option 1 unless you see a need to allow callers to specify the argument as a keyword.

For more details, you might find [the Callables chapter](https://typing.readthedocs.io/en/latest/spec/callables.html) of the Python typing spec to be useful.
